## 1.Load dataSet

In [25]:
import pandas as pd

df = pd.read_csv("dataset.csv")
df.head()

,Id,MSSubClass,MSZoning,LotFrontage,LotArea,Street,Alley,LotShape,LandContour,Utilities,...,PoolArea,PoolQC,Fence,MiscFeature,MiscVal,MoSold,YrSold,SaleType,SaleCondition,SalePrice
0,1,60,RL,65.0,8450,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,NaN,NaN,0,2,2008,WD,Normal,208500
1,2,20,RL,80.0,9600,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,NaN,NaN,0,5,2007,WD,Normal,181500
2,3,60,RL,68.0,11250,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,9,2008,WD,Normal,223500
3,4,70,RL,60.0,9550,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,2,2006,WD,Abnorml,140000
4,5,60,RL,84.0,14260,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,12,2008,WD,Normal,250000


## 2. Separate target and features

In [26]:
X = df.drop("SalePrice", axis=1)
y = df["SalePrice"]

## 3.Split the data

In [30]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

print(f"X_train shape: {X_train.shape}")
print(f"X_test shape: {X_test.shape}")
print(f"y_train shape: {y_train.shape}")
print(f"y_test shape: {y_test.shape}")

X_train shape: (1168, 80)
X_test shape: (292, 80)
y_train shape: (1168,)
y_test shape: (292,)


## 4.Data Understanding

### 4.1 Data Types

#### 4.1.1 - Check data types

In [35]:
# Check the data types of all features
X_train.dtypes

Id                 int64
MSSubClass         int64
MSZoning             str
LotFrontage      float64
LotArea            int64
                  ...   
MiscVal            int64
MoSold             int64
YrSold             int64
SaleType             str
SaleCondition        str
Length: 80, dtype: object

In [36]:
X_train.info()

<class 'pandas.DataFrame'>
Index: 1168 entries, 254 to 1126
Data columns (total 80 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1168 non-null   int64  
 1   MSSubClass     1168 non-null   int64  
 2   MSZoning       1168 non-null   str    
 3   LotFrontage    951 non-null    float64
 4   LotArea        1168 non-null   int64  
 5   Street         1168 non-null   str    
 6   Alley          74 non-null     str    
 7   LotShape       1168 non-null   str    
 8   LandContour    1168 non-null   str    
 9   Utilities      1168 non-null   str    
 10  LotConfig      1168 non-null   str    
 11  LandSlope      1168 non-null   str    
 12  Neighborhood   1168 non-null   str    
 13  Condition1     1168 non-null   str    
 14  Condition2     1168 non-null   str    
 15  BldgType       1168 non-null   str    
 16  HouseStyle     1168 non-null   str    
 17  OverallQual    1168 non-null   int64  
 18  OverallCond    1168 no

#### 4.1.2 - Separate numerical and categorical columns

In [37]:
# Numerical columns
numerical_cols = X_train.select_dtypes(
    include=['int64', 'float64']
).columns

# Categorical columns
categorical_cols = X_train.select_dtypes(
    include=['object']
).columns

print("Number of numerical columns:", len(numerical_cols))
print("Number of categorical columns:", len(categorical_cols))

print("\nNumerical columns:")
print(list(numerical_cols))

print("\nCategorical columns:")
print(list(categorical_cols))

Number of numerical columns: 37
Number of categorical columns: 43

Numerical columns:
['Id', 'MSSubClass', 'LotFrontage', 'LotArea', 'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd', 'MasVnrArea', 'BsmtFinSF1', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', '1stFlrSF', '2ndFlrSF', 'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath', 'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'TotRmsAbvGrd', 'Fireplaces', 'GarageYrBlt', 'GarageCars', 'GarageArea', 'WoodDeckSF', 'OpenPorchSF', 'EnclosedPorch', '3SsnPorch', 'ScreenPorch', 'PoolArea', 'MiscVal', 'MoSold', 'YrSold']

Categorical columns:
['MSZoning', 'Street', 'Alley', 'LotShape', 'LandContour', 'Utilities', 'LotConfig', 'LandSlope', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType', 'HouseStyle', 'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType', 'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2', 'Heating', 'HeatingQC', 'CentralAir'

C:\Users\ASUS\AppData\Local\Temp\ipykernel_7904\2366444756.py:7: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  categorical_cols = X_train.select_dtypes(


#### 4.1.3 - Check unique values of categorical columns

In [38]:
for col in categorical_cols:
    print(f"{col}: {X_train[col].nunique()} unique values")

MSZoning: 5 unique values
Street: 2 unique values
Alley: 2 unique values
LotShape: 4 unique values
LandContour: 4 unique values
Utilities: 2 unique values
LotConfig: 5 unique values
LandSlope: 3 unique values
Neighborhood: 25 unique values
Condition1: 9 unique values
Condition2: 8 unique values
BldgType: 5 unique values
HouseStyle: 8 unique values
RoofStyle: 6 unique values
RoofMatl: 7 unique values
Exterior1st: 15 unique values
Exterior2nd: 16 unique values
MasVnrType: 3 unique values
ExterQual: 4 unique values
ExterCond: 5 unique values
Foundation: 6 unique values
BsmtQual: 4 unique values
BsmtCond: 4 unique values
BsmtExposure: 4 unique values
BsmtFinType1: 6 unique values
BsmtFinType2: 6 unique values
Heating: 6 unique values
HeatingQC: 5 unique values
CentralAir: 2 unique values
Electrical: 4 unique values
KitchenQual: 4 unique values
Functional: 7 unique values
FireplaceQu: 5 unique values
GarageType: 6 unique values
GarageFinish: 3 unique values
GarageQual: 5 unique values
Garag

### 4.2 Missing Value Analysis

#### 4.2.1 - Find missing values

In [39]:
# Count missing values in each column
missing_count = X_train.isnull().sum()

# Show only columns with missing values
missing_count[missing_count > 0].sort_values(ascending=False)

PoolQC          1162
MiscFeature     1122
Alley           1094
Fence            935
MasVnrType       683
FireplaceQu      547
LotFrontage      217
GarageType        64
GarageYrBlt       64
GarageFinish      64
GarageQual        64
GarageCond        64
BsmtCond          28
BsmtFinType1      28
BsmtExposure      28
BsmtQual          28
BsmtFinType2      28
MasVnrArea         6
Electrical         1
dtype: int64

#### 4.2.2 - Calculate missing percentage

In [40]:
missing_percent = (X_train.isnull().sum() / len(X_train)) * 100

missing_table = pd.DataFrame({
    'Missing Count': missing_count,
    'Missing Percentage': missing_percent
})

# Show only columns with missing values
missing_table = missing_table[
    missing_table['Missing Count'] > 0
].sort_values('Missing Percentage', ascending=False)

missing_table

,Missing Count,Missing Percentage
PoolQC,1162,99.486301
MiscFeature,1122,96.061644
Alley,1094,93.664384
Fence,935,80.051370
MasVnrType,683,58.476027
FireplaceQu,547,46.832192
LotFrontage,217,18.578767
GarageType,64,5.479452
GarageYrBlt,64,5.479452
GarageFinish,64,5.479452


#### 4.2.3 - Check whether target has missing values

In [41]:
y_train.isnull().sum()

np.int64(0)

### 4.3 Duplicate Rows

#### 4.3.1 - Count duplicate rows

In [42]:
# Count duplicate rows
duplicate_count = X_train.duplicated().sum()

print("Number of duplicate rows:", duplicate_count)

Number of duplicate rows: 0


### 4.4 Numerical Distribution & Skewness

#### 4.4.1 - Calculate skewness

In [43]:
# Calculate skewness of numerical features
skewness = X_train[numerical_cols].skew()

# Sort from most positively skewed to most negatively skewed
skewness.sort_values(ascending=False)

MiscVal          22.053579
PoolArea         14.396098
LotArea          11.958088
3SsnPorch         9.833911
LowQualFinSF      9.199456
KitchenAbvGr      4.445055
BsmtFinSF2        4.217895
ScreenPorch       4.090393
BsmtHalfBath      4.005786
EnclosedPorch     3.163946
LotFrontage       2.406787
OpenPorchSF       2.331890
MasVnrArea        2.286423
BsmtFinSF1        1.862132
TotalBsmtSF       1.723881
WoodDeckSF        1.587330
MSSubClass        1.438804
GrLivArea         1.425139
1stFlrSF          1.422162
BsmtUnfSF         0.910628
2ndFlrSF          0.801209
OverallCond       0.733180
TotRmsAbvGrd      0.687336
HalfBath          0.651296
Fireplaces        0.637526
BsmtFullBath      0.612032
BedroomAbvGr      0.289882
MoSold            0.241761
OverallQual       0.196575
GarageArea        0.109110
YrSold            0.082593
FullBath          0.038169
Id                0.000920
GarageCars       -0.389732
YearRemodAdd     -0.508067
YearBuilt        -0.623404
GarageYrBlt      -0.666488
d

#### 4.4.2 - Find highly skewed features

In [44]:
highly_skewed = skewness[abs(skewness) > 1]

print("Number of highly skewed features:", len(highly_skewed))
print("\nHighly skewed features:")
print(highly_skewed.sort_values(ascending=False))

Number of highly skewed features: 19

Highly skewed features:
MiscVal          22.053579
PoolArea         14.396098
LotArea          11.958088
3SsnPorch         9.833911
LowQualFinSF      9.199456
KitchenAbvGr      4.445055
BsmtFinSF2        4.217895
ScreenPorch       4.090393
BsmtHalfBath      4.005786
EnclosedPorch     3.163946
LotFrontage       2.406787
OpenPorchSF       2.331890
MasVnrArea        2.286423
BsmtFinSF1        1.862132
TotalBsmtSF       1.723881
WoodDeckSF        1.587330
MSSubClass        1.438804
GrLivArea         1.425139
1stFlrSF          1.422162
dtype: float64


### 4.5 Categorical Features

#### 4.5.1 - Check category counts

In [45]:
for col in categorical_cols:
    print(f"\n===== {col} =====")
    print(X_train[col].value_counts(dropna=False))


===== MSZoning =====
MSZoning
RL         924
RM         172
FV          53
RH          15
C (all)      4
Name: count, dtype: int64

===== Street =====
Street
Pave    1164
Grvl       4
Name: count, dtype: int64

===== Alley =====
Alley
NaN     1094
Grvl      44
Pave      30
Name: count, dtype: int64

===== LotShape =====
LotShape
Reg    729
IR1    394
IR2     37
IR3      8
Name: count, dtype: int64

===== LandContour =====
LandContour
Lvl    1059
Bnk      48
HLS      35
Low      26
Name: count, dtype: int64

===== Utilities =====
Utilities
AllPub    1167
NoSeWa       1
Name: count, dtype: int64

===== LotConfig =====
LotConfig
Inside     822
Corner     221
CulDSac     84
FR2         38
FR3          3
Name: count, dtype: int64

===== LandSlope =====
LandSlope
Gtl    1108
Mod      51
Sev       9
Name: count, dtype: int64

===== Neighborhood =====
Neighborhood
NAmes      181
CollgCr    115
OldTown     91
Edwards     87
Somerst     69
NWAmes      66
Gilbert     65
NridgHt     61
Sawyer    

#### 4.5.2 - Check unique categories

In [46]:
for col in categorical_cols:
    print(f"{col}:")
    print(X_train[col].unique())

MSZoning:
<StringArray>
['RL', 'RM', 'FV', 'RH', 'C (all)']
Length: 5, dtype: str
Street:
<StringArray>
['Pave', 'Grvl']
Length: 2, dtype: str
Alley:
<StringArray>
[nan, 'Pave', 'Grvl']
Length: 3, dtype: str
LotShape:
<StringArray>
['Reg', 'IR1', 'IR2', 'IR3']
Length: 4, dtype: str
LandContour:
<StringArray>
['Lvl', 'Bnk', 'HLS', 'Low']
Length: 4, dtype: str
Utilities:
<StringArray>
['AllPub', 'NoSeWa']
Length: 2, dtype: str
LotConfig:
<StringArray>
['Inside', 'Corner', 'CulDSac', 'FR2', 'FR3']
Length: 5, dtype: str
LandSlope:
<StringArray>
['Gtl', 'Mod', 'Sev']
Length: 3, dtype: str
Neighborhood:
<StringArray>
[  'NAmes', 'Gilbert', 'Edwards',   'SWISU', 'CollgCr',  'IDOTRR',  'NWAmes',
 'Somerst',  'Sawyer', 'Crawfor', 'OldTown',  'Timber', 'BrkSide', 'SawyerW',
 'NridgHt', 'Mitchel', 'NoRidge', 'Blmngtn',  'BrDale', 'NPkVill', 'Veenker',
 'ClearCr', 'StoneBr', 'MeadowV', 'Blueste']
Length: 25, dtype: str
Condition1:
<StringArray>
['Norm', 'Feedr', 'RRAn', 'Artery', 'PosA', 'PosN', '

### 4.6 Outlier Analysis

#### 4.6.1 - Use IQR method

In [47]:
import pandas as pd

outlier_results = []

for col in numerical_cols:
    Q1 = X_train[col].quantile(0.25)
    Q3 = X_train[col].quantile(0.75)
    IQR = Q3 - Q1

    lower = Q1 - 1.5 * IQR
    upper = Q3 + 1.5 * IQR

    outlier_count = ((X_train[col] < lower) | 
                     (X_train[col] > upper)).sum()

    outlier_results.append({
        'Feature': col,
        'Q1': Q1,
        'Q3': Q3,
        'IQR': IQR,
        'Lower Bound': lower,
        'Upper Bound': upper,
        'Outlier Count': outlier_count
    })

outlier_table = pd.DataFrame(outlier_results)

outlier_table = outlier_table.sort_values(
    'Outlier Count',
    ascending=False
)

outlier_table

,Feature,Q1,Q3,IQR,Lower Bound,Upper Bound,Outlier Count
30,EnclosedPorch,0.00,0.00,0.00,0.000,0.000,160
10,BsmtFinSF2,0.00,0.00,0.00,0.000,0.000,131
5,OverallCond,5.00,6.00,1.00,3.500,7.500,99
32,ScreenPorch,0.00,0.00,0.00,0.000,0.000,97
1,MSSubClass,20.00,70.00,50.00,-55.000,145.000,84
8,MasVnrArea,0.00,166.00,166.00,-249.000,415.000,77
2,LotFrontage,59.00,80.00,21.00,27.500,111.500,68
18,BsmtHalfBath,0.00,0.00,0.00,0.000,0.000,66
29,OpenPorchSF,0.00,74.00,74.00,-111.000,185.000,60
22,KitchenAbvGr,1.00,1.00,0.00,1.000,1.000,56


#### 4.6.2 - Show only features with outliers

In [48]:
outlier_table[outlier_table['Outlier Count'] > 0]

,Feature,Q1,Q3,IQR,Lower Bound,Upper Bound,Outlier Count
30,EnclosedPorch,0.00,0.00,0.00,0.000,0.000,160
10,BsmtFinSF2,0.00,0.00,0.00,0.000,0.000,131
5,OverallCond,5.00,6.00,1.00,3.500,7.500,99
32,ScreenPorch,0.00,0.00,0.00,0.000,0.000,97
1,MSSubClass,20.00,70.00,50.00,-55.000,145.000,84
8,MasVnrArea,0.00,166.00,166.00,-249.000,415.000,77
2,LotFrontage,59.00,80.00,21.00,27.500,111.500,68
18,BsmtHalfBath,0.00,0.00,0.00,0.000,0.000,66
29,OpenPorchSF,0.00,74.00,74.00,-111.000,185.000,60
22,KitchenAbvGr,1.00,1.00,0.00,1.000,1.000,56


#### 4.6.3 Calculate outlier percentage

In [49]:
outlier_table['Outlier Percentage'] = (
    outlier_table['Outlier Count'] / len(X_train)
) * 100

outlier_table[
    outlier_table['Outlier Count'] > 0
][['Feature', 'Outlier Count', 'Outlier Percentage']]

,Feature,Outlier Count,Outlier Percentage
30,EnclosedPorch,160,13.698630
10,BsmtFinSF2,131,11.215753
5,OverallCond,99,8.476027
32,ScreenPorch,97,8.304795
1,MSSubClass,84,7.191781
8,MasVnrArea,77,6.592466
2,LotFrontage,68,5.821918
18,BsmtHalfBath,66,5.650685
29,OpenPorchSF,60,5.136986
22,KitchenAbvGr,56,4.794521


## 5.Missing Value Handling

### 5.1 Categorical Missing Value Handling

#### 5.1.1 Apply "None"

In [50]:
none_cols = [
    'PoolQC',
    'MiscFeature',
    'Alley',
    'Fence',
    'MasVnrType',
    'FireplaceQu',
    'GarageType',
    'GarageFinish',
    'GarageQual',
    'GarageCond',
    'BsmtCond',
    'BsmtFinType1',
    'BsmtExposure',
    'BsmtQual',
    'BsmtFinType2'
]

X_train[none_cols] = X_train[none_cols].fillna('None')
X_test[none_cols] = X_test[none_cols].fillna('None')

In [51]:
X_train[none_cols].isnull().sum()

PoolQC          0
MiscFeature     0
Alley           0
Fence           0
MasVnrType      0
FireplaceQu     0
GarageType      0
GarageFinish    0
GarageQual      0
GarageCond      0
BsmtCond        0
BsmtFinType1    0
BsmtExposure    0
BsmtQual        0
BsmtFinType2    0
dtype: int64

In [52]:
X_test[none_cols].isnull().sum()

PoolQC          0
MiscFeature     0
Alley           0
Fence           0
MasVnrType      0
FireplaceQu     0
GarageType      0
GarageFinish    0
GarageQual      0
GarageCond      0
BsmtCond        0
BsmtFinType1    0
BsmtExposure    0
BsmtQual        0
BsmtFinType2    0
dtype: int64

### 5.2 Numerical Missing Value Handling

#### 5.2.1 LotFrontage → Median

In [53]:
lotfrontage_median = X_train['LotFrontage'].median()

X_train['LotFrontage'] = X_train['LotFrontage'].fillna(
    lotfrontage_median
)

X_test['LotFrontage'] = X_test['LotFrontage'].fillna(
    lotfrontage_median
)

#### 5.2.2 GarageYrBlt → Handle based on garage absence

In [54]:
X_train['GarageYrBlt'] = X_train['GarageYrBlt'].fillna(0)
X_test['GarageYrBlt'] = X_test['GarageYrBlt'].fillna(0)

#### 5.2.3 MasVnrArea → 0 when no veneer

In [55]:
X_train['MasVnrArea'] = X_train['MasVnrArea'].fillna(0)
X_test['MasVnrArea'] = X_test['MasVnrArea'].fillna(0)

#### 5.2.4 Electrical → Mode imputation

In [57]:
electrical_mode = X_train['Electrical'].mode()[0]

X_train['Electrical'] = X_train['Electrical'].fillna(electrical_mode)
X_test['Electrical'] = X_test['Electrical'].fillna(electrical_mode)

#### 5.2.5 Verify all missing values

In [58]:
print("Missing values in X_train:")
print(X_train.isnull().sum()[X_train.isnull().sum() > 0])

print("\nMissing values in X_test:")
print(X_test.isnull().sum()[X_test.isnull().sum() > 0])

Missing values in X_train:
Series([], dtype: int64)

Missing values in X_test:
Series([], dtype: int64)


## 6. Outlier Handling

## 7. Feature Engineering

Potential outliers were identified using the IQR method during the outlier analysis.

However, the detected outliers were not automatically removed because:

- Many extreme values represent legitimate house characteristics.
- Several features are zero-inflated, where most values are 0. For example:
  - EnclosedPorch
  - ScreenPorch
  - BsmtFinSF2
  - PoolArea
  - 3SsnPorch
  - MiscVal
- Some features are discrete or coded variables, such as:
  - MSSubClass
  - OverallQual
  - OverallCond
  - KitchenAbvGr
  - GarageCars
- Removing observations only because they fall outside the IQR bounds could remove valid information from the dataset.

Therefore, no rows were removed based solely on the IQR outlier detection.

Extreme values in highly skewed numerical features will instead be considered during **Feature Engineering**, where appropriate transformations may be applied.

### 7.1 Total House Area (`TotalSF`)

The total area across the basement, first floor, and second floor was combined into a single feature.

**Formula:**

`TotalSF = TotalBsmtSF + 1stFlrSF + 2ndFlrSF`

In [59]:
X_train['TotalSF'] = (
    X_train['TotalBsmtSF'] +
    X_train['1stFlrSF'] +
    X_train['2ndFlrSF']
)

X_test['TotalSF'] = (
    X_test['TotalBsmtSF'] +
    X_test['1stFlrSF'] +
    X_test['2ndFlrSF']
)

### 7.2 Total Bathrooms (`TotalBathrooms`)

The total number of bathrooms was calculated by combining full and half bathrooms. Half bathrooms were counted as 0.5 because they provide fewer facilities than full bathrooms.

**Formula:**

`TotalBathrooms = FullBath + 0.5 × HalfBath + BsmtFullBath + 0.5 × BsmtHalfBath`

In [60]:
X_train['TotalBathrooms'] = (
    X_train['FullBath']
    + 0.5 * X_train['HalfBath']
    + X_train['BsmtFullBath']
    + 0.5 * X_train['BsmtHalfBath']
)

X_test['TotalBathrooms'] = (
    X_test['FullBath']
    + 0.5 * X_test['HalfBath']
    + X_test['BsmtFullBath']
    + 0.5 * X_test['BsmtHalfBath']
)

### 7.3 Total Porch Area (`TotalPorchSF`)

The areas of different porch and deck types were combined into a single feature representing the total outdoor porch/deck area.

**Formula:**

`TotalPorchSF = OpenPorchSF + 3SsnPorch + EnclosedPorch + ScreenPorch + WoodDeckSF`

In [61]:
X_train['TotalPorchSF'] = (
    X_train['OpenPorchSF']
    + X_train['3SsnPorch']
    + X_train['EnclosedPorch']
    + X_train['ScreenPorch']
    + X_train['WoodDeckSF']
)

X_test['TotalPorchSF'] = (
    X_test['OpenPorchSF']
    + X_test['3SsnPorch']
    + X_test['EnclosedPorch']
    + X_test['ScreenPorch']
    + X_test['WoodDeckSF']
)

### 7.4 House Age (`HouseAge`)

The age of the house at the time it was sold was calculated using the year the house was sold and the year it was originally built.

**Formula:**

`HouseAge = YrSold - YearBuilt`

In [62]:
X_train['HouseAge'] = X_train['YrSold'] - X_train['YearBuilt']
X_test['HouseAge'] = X_test['YrSold'] - X_test['YearBuilt']

### 7.5 Years Since Remodeling (`YearsSinceRemod`)

The number of years since the house was last remodeled was calculated using the year of sale and the year of the most recent remodeling.

**Formula:**

`YearsSinceRemod = YrSold - YearRemodAdd`

In [63]:
X_train['YearsSinceRemod'] = (
    X_train['YrSold'] - X_train['YearRemodAdd']
)

X_test['YearsSinceRemod'] = (
    X_test['YrSold'] - X_test['YearRemodAdd']
)

### 7.6 Total Finished Basement Area (`TotalBsmtFinSF`)

The finished basement areas from both basement finishing types were combined into a single feature.

**Formula:**

`TotalBsmtFinSF = BsmtFinSF1 + BsmtFinSF2`

In [64]:
X_train['TotalBsmtFinSF'] = (
    X_train['BsmtFinSF1'] +
    X_train['BsmtFinSF2']
)

X_test['TotalBsmtFinSF'] = (
    X_test['BsmtFinSF1'] +
    X_test['BsmtFinSF2']
)

### 7.7 Total Living Area (`TotalLivingArea`)

The above-ground living area and basement area were combined to represent the total living space of the house.

**Formula:**

`TotalLivingArea = GrLivArea + TotalBsmtSF`

In [65]:
X_train['TotalLivingArea'] = (
    X_train['GrLivArea'] +
    X_train['TotalBsmtSF']
)

X_test['TotalLivingArea'] = (
    X_test['GrLivArea'] +
    X_test['TotalBsmtSF']
)

### 7.8 Check the New Features

In [66]:
new_features = [
    'TotalSF',
    'TotalBathrooms',
    'TotalPorchSF',
    'HouseAge',
    'YearsSinceRemod',
    'TotalBsmtFinSF',
    'TotalLivingArea'
]

X_train[new_features].describe()

,TotalSF,TotalBathrooms,TotalPorchSF,HouseAge,YearsSinceRemod,TotalBsmtFinSF,TotalLivingArea
count,1168.000000,1168.000000,1168.000000,1168.00000,1168.000000,1168.000000,1168.000000
mean,2583.476884,2.222175,186.584760,36.85274,22.921233,491.176370,2589.172945
std,821.305316,0.779502,161.131777,30.70311,20.714585,478.996643,823.125867
min,334.000000,1.000000,0.000000,0.00000,-1.000000,0.000000,334.000000
25%,2032.500000,2.000000,45.000000,7.00000,4.000000,0.000000,2041.000000
50%,2494.000000,2.000000,168.000000,35.00000,14.000000,482.500000,2500.000000
75%,3013.500000,2.500000,275.250000,55.00000,42.000000,803.250000,3028.250000
max,11752.000000,6.000000,1027.000000,136.00000,60.000000,5644.000000,11752.000000


In [67]:
X_test[new_features].describe()

,TotalSF,TotalBathrooms,TotalPorchSF,HouseAge,YearsSinceRemod,TotalBsmtFinSF,TotalLivingArea
count,292.000000,292.000000,292.000000,292.000000,292.000000,292.000000,292.000000
mean,2501.335616,2.164384,160.308219,35.328767,23.065068,486.239726,2507.773973
std,821.470760,0.808268,135.525910,28.383629,20.377010,465.136063,823.679406
min,827.000000,1.000000,0.000000,0.000000,0.000000,0.000000,827.000000
25%,1941.750000,1.500000,43.750000,9.000000,4.000000,0.000000,1947.250000
50%,2393.000000,2.000000,144.000000,34.000000,15.000000,453.000000,2398.000000
75%,2938.500000,2.500000,240.000000,54.000000,39.000000,769.500000,2946.250000
max,6760.000000,4.000000,678.000000,127.000000,60.000000,2188.000000,6760.000000


## 8. Feature Selection

### 8.1 Remove Id

In [69]:
X_train = X_train.drop(columns=['Id'])
X_test = X_test.drop(columns=['Id'])

In [70]:
print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)

X_train shape: (1168, 86)
X_test shape: (292, 86)


### 8.2 Correlation Analysis

In [71]:
# Create a temporary dataframe with numerical features and target
corr_data = X_train.copy()
corr_data['SalePrice'] = y_train

# Calculate correlations with SalePrice
correlation = corr_data.corr(numeric_only=True)['SalePrice'].sort_values(
    ascending=False
)

print(correlation)

SalePrice          1.000000
OverallQual        0.785555
TotalSF            0.765586
TotalLivingArea    0.763242
GrLivArea          0.695652
GarageCars         0.640991
GarageArea         0.624139
TotalBathrooms     0.621280
TotalBsmtSF        0.597766
1stFlrSF           0.587883
FullBath           0.552546
TotRmsAbvGrd       0.520388
YearBuilt          0.516501
YearRemodAdd       0.508593
MasVnrArea         0.457608
Fireplaces         0.457549
TotalPorchSF       0.388528
BsmtFinSF1         0.359460
TotalBsmtFinSF     0.342614
WoodDeckSF         0.329843
2ndFlrSF           0.314030
LotFrontage        0.312157
OpenPorchSF        0.299969
HalfBath           0.280481
GarageYrBlt        0.272056
LotArea            0.266204
BsmtFullBath       0.226346
BsmtUnfSF          0.222487
BedroomAbvGr       0.156211
ScreenPorch        0.119172
PoolArea           0.115630
3SsnPorch          0.051532
MoSold             0.041890
BsmtFinSF2        -0.005731
YrSold            -0.009099
LowQualFinSF      -0

### 8.3 Check Highly Correlated / Redundant Features

In [72]:
# Calculate correlation matrix for numerical features
numeric_corr = X_train.corr(numeric_only=True)

# Find pairs of features with correlation greater than 0.80
high_corr_pairs = []

for i in range(len(numeric_corr.columns)):
    for j in range(i + 1, len(numeric_corr.columns)):
        corr_value = numeric_corr.iloc[i, j]
        
        if abs(corr_value) > 0.80:
            high_corr_pairs.append((
                numeric_corr.columns[i],
                numeric_corr.columns[j],
                corr_value
            ))

# Display highly correlated feature pairs
for feature1, feature2, corr_value in high_corr_pairs:
    print(f"{feature1} <-> {feature2}: {corr_value:.3f}")

YearBuilt <-> HouseAge: -0.999
YearRemodAdd <-> YearsSinceRemod: -0.998
BsmtFinSF1 <-> TotalBsmtFinSF: 0.944
TotalBsmtSF <-> 1stFlrSF: 0.827
TotalBsmtSF <-> TotalSF: 0.827
TotalBsmtSF <-> TotalLivingArea: 0.823
1stFlrSF <-> TotalSF: 0.801
GrLivArea <-> TotRmsAbvGrd: 0.821
GrLivArea <-> TotalSF: 0.872
GrLivArea <-> TotalLivingArea: 0.878
GarageCars <-> GarageArea: 0.883
TotalSF <-> TotalLivingArea: 0.998


### 8.4 Remove Redundant Features

Highly correlated features were identified in the correlation analysis. When two features provide very similar information, one of them can be removed to reduce redundancy.

**Features removed:**

- `YearBuilt` — highly correlated with `HouseAge` (-0.999). `HouseAge` represents the age of the house at the time of sale.
- `YearRemodAdd` — highly correlated with `YearsSinceRemod` (-0.998). `YearsSinceRemod` represents the number of years since the house was remodeled.
- `BsmtFinSF1` — highly correlated with `TotalBsmtFinSF` (0.944). `TotalBsmtFinSF` provides the combined finished basement area.
- `TotalLivingArea` — almost perfectly correlated with `TotalSF` (0.998). `TotalSF` provides the combined area of the basement, first floor, and second floor.
- `GarageArea` — highly correlated with `GarageCars` (0.883). `GarageCars` represents the garage capacity in terms of the number of cars.

**Features removed:**

`YearBuilt`, `YearRemodAdd`, `BsmtFinSF1`, `TotalLivingArea`, `GarageArea`

The remaining features, including the newly engineered features, will be used for the next stages of preprocessing and model development.

In [73]:
features_to_remove = [
    'YearBuilt',
    'YearRemodAdd',
    'BsmtFinSF1',
    'TotalLivingArea',
    'GarageArea'
]

X_train = X_train.drop(columns=features_to_remove)
X_test = X_test.drop(columns=features_to_remove)

In [74]:
print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)

X_train shape: (1168, 81)
X_test shape: (292, 81)


### 8.5 Final Feature Selection Check

After removing the redundant features, the final feature set was checked to ensure that the selected features are present in both the training and testing datasets.

The following features were removed during feature selection:

- `Id`
- `YearBuilt`
- `YearRemodAdd`
- `BsmtFinSF1`
- `TotalLivingArea`
- `GarageArea`

The remaining features include the useful original features together with the engineered features created in Step 7.

The training and testing datasets were kept consistent so that both contain exactly the same predictor features.


In [75]:
print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)

print("\nFeatures in X_train:")
print(X_train.columns.tolist())

print("\nFeatures in X_test:")
print(X_test.columns.tolist())

print("\nSame features in train and test:",
      list(X_train.columns) == list(X_test.columns))

X_train shape: (1168, 81)
X_test shape: (292, 81)

Features in X_train:
['MSSubClass', 'MSZoning', 'LotFrontage', 'LotArea', 'Street', 'Alley', 'LotShape', 'LandContour', 'Utilities', 'LotConfig', 'LandSlope', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType', 'HouseStyle', 'OverallQual', 'OverallCond', 'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType', 'MasVnrArea', 'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', 'Heating', 'HeatingQC', 'CentralAir', 'Electrical', '1stFlrSF', '2ndFlrSF', 'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath', 'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'KitchenQual', 'TotRmsAbvGrd', 'Functional', 'Fireplaces', 'FireplaceQu', 'GarageType', 'GarageYrBlt', 'GarageFinish', 'GarageCars', 'GarageQual', 'GarageCond', 'PavedDrive', 'WoodDeckSF', 'OpenPorchSF', 'EnclosedPorch', '3SsnPorch', 'ScreenPorch', 'PoolArea',

## 9. Categorical Encoding

### 9.1 Identify Categorical Features

In [76]:
categorical_cols = X_train.select_dtypes(
    include=['object']
).columns.tolist()

print("Number of categorical features:", len(categorical_cols))
print(categorical_cols)

Number of categorical features: 43
['MSZoning', 'Street', 'Alley', 'LotShape', 'LandContour', 'Utilities', 'LotConfig', 'LandSlope', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType', 'HouseStyle', 'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType', 'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2', 'Heating', 'HeatingQC', 'CentralAir', 'Electrical', 'KitchenQual', 'Functional', 'FireplaceQu', 'GarageType', 'GarageFinish', 'GarageQual', 'GarageCond', 'PavedDrive', 'PoolQC', 'Fence', 'MiscFeature', 'SaleType', 'SaleCondition']


C:\Users\ASUS\AppData\Local\Temp\ipykernel_7904\170624432.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  categorical_cols = X_train.select_dtypes(


### 9.2 Identify Numerical Features

In [77]:
numerical_cols = X_train.select_dtypes(
    include=['int64', 'float64']
).columns.tolist()

print("Number of numerical features:", len(numerical_cols))

Number of numerical features: 38


### 9.3 Identify Ordinal and Nominal Categorical Features

The categorical features were divided into **ordinal** and **nominal** features based on whether their categories have a meaningful order.

There are **43 categorical features** in total:

- **18 ordinal features**
- **25 nominal features**

Therefore:

`18 + 25 = 43`

The classification was verified to ensure that every categorical feature belongs to exactly one of these two groups.

**Ordinal categorical features:**

`ExterQual`, `ExterCond`, `BsmtQual`, `BsmtCond`, `HeatingQC`, `KitchenQual`, `FireplaceQu`, `GarageQual`, `GarageCond`, `BsmtExposure`, `BsmtFinType1`, `BsmtFinType2`, `GarageFinish`, `Functional`, `PavedDrive`, `LotShape`, `LandSlope`, `PoolQC`

**Nominal categorical features:**

`MSZoning`, `Street`, `Alley`, `LandContour`, `LotConfig`, `Neighborhood`, `Condition1`, `Condition2`, `BldgType`, `HouseStyle`, `RoofStyle`, `RoofMatl`, `Exterior1st`, `Exterior2nd`, `MasVnrType`, `Foundation`, `Heating`, `CentralAir`, `Electrical`, `GarageType`, `SaleType`, `SaleCondition`, `MiscFeature`, `Fence`, `Utilities`

Ordinal features will be encoded while preserving their category order, whereas nominal features will be encoded without imposing an artificial order.

In [84]:
# Ordinal categorical features
ordinal_cols = [
    'ExterQual',
    'ExterCond',
    'BsmtQual',
    'BsmtCond',
    'HeatingQC',
    'KitchenQual',
    'FireplaceQu',
    'GarageQual',
    'GarageCond',
    'BsmtExposure',
    'BsmtFinType1',
    'BsmtFinType2',
    'GarageFinish',
    'Functional',
    'PavedDrive',
    'LotShape',
    'LandSlope',
    'PoolQC'
]

# Nominal categorical features
nominal_cols = [
    'MSZoning',
    'Street',
    'Alley',
    'LandContour',
    'LotConfig',
    'Neighborhood',
    'Condition1',
    'Condition2',
    'BldgType',
    'HouseStyle',
    'RoofStyle',
    'RoofMatl',
    'Exterior1st',
    'Exterior2nd',
    'MasVnrType',
    'Foundation',
    'Heating',
    'CentralAir',
    'Electrical',
    'GarageType',
    'SaleType',
    'SaleCondition',
    'MiscFeature',
    'Fence',
    'Utilities'
]

print("Ordinal features:", len(ordinal_cols))
print(ordinal_cols)

print("\nNominal features:", len(nominal_cols))
print(nominal_cols)

Ordinal features: 18
['ExterQual', 'ExterCond', 'BsmtQual', 'BsmtCond', 'HeatingQC', 'KitchenQual', 'FireplaceQu', 'GarageQual', 'GarageCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2', 'GarageFinish', 'Functional', 'PavedDrive', 'LotShape', 'LandSlope', 'PoolQC']

Nominal features: 25
['MSZoning', 'Street', 'Alley', 'LandContour', 'LotConfig', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType', 'HouseStyle', 'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType', 'Foundation', 'Heating', 'CentralAir', 'Electrical', 'GarageType', 'SaleType', 'SaleCondition', 'MiscFeature', 'Fence', 'Utilities']


In [85]:
print("Total categorical features:", len(categorical_cols))
print("Total ordinal features:", len(ordinal_cols))
print("Total nominal features:", len(nominal_cols))

print("\nCheck:")
print("Ordinal + Nominal =", len(ordinal_cols) + len(nominal_cols))
print("Equal to categorical =", 
      len(ordinal_cols) + len(nominal_cols) == len(categorical_cols))

Total categorical features: 43
Total ordinal features: 18
Total nominal features: 25

Check:
Ordinal + Nominal = 43
Equal to categorical = True


### 9.4 Ordinal Encoding

The 18 ordinal categorical features were encoded using **Ordinal Encoding** because their categories have a meaningful order.

For example:

`ExterQual: Po < Fa < TA < Gd < Ex`

The category order was explicitly defined for each ordinal feature so that the encoding preserves the original ranking.

`fit_transform()` was applied to the training data, while `transform()` was applied to the test data using the same learned encoding rules. This prevents data leakage.


In [87]:
from sklearn.preprocessing import OrdinalEncoder

ordinal_categories = [
    ['None', 'Po', 'Fa', 'TA', 'Gd', 'Ex'],  # ExterQual
    ['None', 'Po', 'Fa', 'TA', 'Gd', 'Ex'],  # ExterCond
    ['None', 'Po', 'Fa', 'TA', 'Gd', 'Ex'],  # BsmtQual
    ['None', 'Po', 'Fa', 'TA', 'Gd', 'Ex'],  # BsmtCond
    ['None', 'Po', 'Fa', 'TA', 'Gd', 'Ex'],  # HeatingQC
    ['None', 'Po', 'Fa', 'TA', 'Gd', 'Ex'],  # KitchenQual
    ['None', 'Po', 'Fa', 'TA', 'Gd', 'Ex'],  # FireplaceQu
    ['None', 'Po', 'Fa', 'TA', 'Gd', 'Ex'],  # GarageQual
    ['None', 'Po', 'Fa', 'TA', 'Gd', 'Ex'],  # GarageCond
    ['None', 'No', 'Mn', 'Av', 'Gd'],       # BsmtExposure
    ['None', 'Unf', 'LwQ', 'Rec', 'BLQ', 'ALQ', 'GLQ'],  # BsmtFinType1
    ['None', 'Unf', 'LwQ', 'Rec', 'BLQ', 'ALQ', 'GLQ'],  # BsmtFinType2
    ['None', 'Unf', 'RFn', 'Fin'],           # GarageFinish
    ['Sal', 'Sev', 'Maj2', 'Maj1', 'Mod', 'Min2', 'Min1', 'Typ'],  # Functional
    ['N', 'P', 'Y'],                          # PavedDrive
    ['IR3', 'IR2', 'IR1', 'Reg'],            # LotShape
    ['Sev', 'Mod', 'Gtl'],                    # LandSlope
    ['None', 'Fa', 'TA', 'Gd', 'Ex']         # PoolQC
]

ordinal_encoder = OrdinalEncoder(
    categories=ordinal_categories,
    handle_unknown='use_encoded_value',
    unknown_value=-1
)

X_train_ordinal = ordinal_encoder.fit_transform(
    X_train[ordinal_cols]
)

X_test_ordinal = ordinal_encoder.transform(
    X_test[ordinal_cols]
)

### 9.5 One-Hot Encoding for Nominal Features

The 25 nominal categorical features were encoded using **One-Hot Encoding** because their categories do not have a meaningful order.

Each category was converted into a separate binary feature containing `0` or `1`.

`handle_unknown='ignore'` was used so that unseen categories in the test data do not cause an encoding error.

`fit_transform()` was applied to the training data, while `transform()` was applied to the test data using the same encoding rules. This prevents data leakage.

In [88]:
from sklearn.preprocessing import OneHotEncoder

nominal_encoder = OneHotEncoder(
    handle_unknown='ignore',
    sparse_output=False
)

X_train_nominal = nominal_encoder.fit_transform(
    X_train[nominal_cols]
)

X_test_nominal = nominal_encoder.transform(
    X_test[nominal_cols]
)

In [89]:
print("Encoded nominal X_train shape:", X_train_nominal.shape)
print("Encoded nominal X_test shape:", X_test_nominal.shape)

Encoded nominal X_train shape: (1168, 174)
Encoded nominal X_test shape: (292, 174)


### 9.6 Combine Numerical, Ordinal, and Nominal Features

In [90]:
import numpy as np

# Select numerical features
X_train_numeric = X_train[numerical_cols].values
X_test_numeric = X_test[numerical_cols].values

# Combine all encoded and numerical features
X_train_encoded = np.hstack([
    X_train_numeric,
    X_train_ordinal,
    X_train_nominal
])

X_test_encoded = np.hstack([
    X_test_numeric,
    X_test_ordinal,
    X_test_nominal
])

print("Final X_train shape:", X_train_encoded.shape)
print("Final X_test shape:", X_test_encoded.shape)

Final X_train shape: (1168, 230)
Final X_test shape: (292, 230)


### 9.7 Encoding Verification

After encoding the categorical features, the numerical, ordinal, and nominal features were successfully combined.

The final datasets contain the same number of features:

- Training data: **(1168, 230)**
- Testing data: **(292, 230)**

The final feature count is **230**, including the original numerical features, ordinal-encoded features, and one-hot encoded nominal features.

The training and testing datasets have matching feature dimensions and are ready for feature scaling.

## 10. Feature Scaling

### 10.1 Standard Scaling

Standard scaling was applied to the final encoded features to bring numerical features with different ranges to a comparable scale.

The **StandardScaler** transforms each feature using its mean and standard deviation.

**Formula:**

`z = (x - mean) / standard deviation`

After scaling, each feature is approximately centered around **0** with a standard deviation of **1**.

In [97]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_train_scaled = scaler.fit_transform(X_train_encoded)
X_test_scaled = scaler.transform(X_test_encoded)

### 10.2 Verify Scaling

In [98]:
print("Scaled X_train shape:", X_train_scaled.shape)
print("Scaled X_test shape:", X_test_scaled.shape)

Scaled X_train shape: (1168, 230)
Scaled X_test shape: (292, 230)


# House Price Prediction - Preprocessing & Feature Engineering Summary

## 1. Load the Dataset

The house price dataset was loaded and separated into input features (`X`) and the target variable (`y`).

- `X` → house features
- `y` → `SalePrice`

---

## 2. Separate Features and Target

The target variable was separated from the input features to prevent the target from being used as a predictor.

`X = Features`

`y = SalePrice`

---

## 3. Train-Test Split

The original training dataset was divided into training and testing sets.

- Training set: **1168 rows**
- Testing set: **292 rows**

The training set was used to learn preprocessing rules and train models, while the testing set was kept for evaluation.

---

## 4. Data Understanding and Analysis

### 4.1 Data Types

The dataset contains both:

- Numerical features
- Categorical features

Categorical features were identified using their data types.

---

### 4.2 Missing Value Analysis

Missing values were analyzed for all features.

Several categorical features had missing values because the corresponding house feature was absent rather than unknown.

Examples:

- `PoolQC`
- `MiscFeature`
- `Alley`
- `Fence`
- `GarageType`
- `BsmtQual`

---

### 4.3 Duplicate Analysis

Duplicate rows were checked using:

`X_train.duplicated().sum()`

No exact duplicate rows were found.

---

### 4.4 Numerical Distribution and Skewness

The numerical features were analyzed for their distributions and skewness.

Skewness was examined to identify numerical features with strongly asymmetric distributions that may require transformation during feature engineering.

---

### 4.5 Categorical Feature Analysis

The categorical features and their unique categories were analyzed.

The dataset contains **43 categorical features**.

The categorical features were later divided into:

- **18 ordinal features**
- **25 nominal features**

---

### 4.6 Outlier Analysis

The IQR method was used to identify potential numerical outliers.

However, the detected outliers were not automatically removed because many extreme values represent legitimate house characteristics.

Several features were also zero-inflated, meaning that many observations had a value of zero.

Therefore, no rows were removed based only on IQR outlier detection.

---

## 5. Missing Value Handling

Missing values were handled according to the meaning of each feature.

### Categorical Features

Missing values representing the absence of a house feature were replaced with:

`None`

### Numerical Features

- `LotFrontage` → median imputation
- `GarageYrBlt` → `0`
- `MasVnrArea` → `0`

### `Electrical`

`Electrical` is a categorical feature, so its missing value was replaced using the mode.

After preprocessing:

- Missing values in `X_train` → **0**
- Missing values in `X_test` → **0**

---

## 6. Outlier Handling

Potential outliers were identified using the IQR method.

The outliers were **not automatically removed** because:

- Extreme values can represent legitimate houses.
- Several variables are zero-inflated.
- Some variables are discrete or coded values.
- Removing observations could remove useful information.

Therefore:

**No rows were removed based solely on IQR outlier detection.**

---

## 7. Feature Engineering

New numerical features were created by combining existing features into meaningful house-related measurements.

### 7.1 Total House Area (`TotalSF`)

The total area across the basement, first floor, and second floor was combined into a single feature.

**Formula:**

`TotalSF = TotalBsmtSF + 1stFlrSF + 2ndFlrSF`

### 7.2 Total Bathrooms (`TotalBathrooms`)

Full and half bathrooms were combined into a single feature. Half bathrooms were counted as 0.5.

**Formula:**

`TotalBathrooms = FullBath + 0.5 × HalfBath + BsmtFullBath + 0.5 × BsmtHalfBath`

### 7.3 Total Porch Area (`TotalPorchSF`)

Different porch and deck areas were combined into one feature.

**Formula:**

`TotalPorchSF = OpenPorchSF + 3SsnPorch + EnclosedPorch + ScreenPorch + WoodDeckSF`

### 7.4 House Age (`HouseAge`)

The age of the house at the time of sale was calculated.

**Formula:**

`HouseAge = YrSold - YearBuilt`

### 7.5 Years Since Remodeling (`YearsSinceRemod`)

The number of years since the house was remodeled was calculated.

**Formula:**

`YearsSinceRemod = YrSold - YearRemodAdd`

### 7.6 Total Finished Basement Area (`TotalBsmtFinSF`)

The finished areas of both basement finishing types were combined.

**Formula:**

`TotalBsmtFinSF = BsmtFinSF1 + BsmtFinSF2`

### 7.7 Total Living Area (`TotalLivingArea`)

The above-ground living area and basement area were combined.

**Formula:**

`TotalLivingArea = GrLivArea + TotalBsmtSF`

---

## 8. Feature Selection

Feature selection was performed to remove unnecessary or redundant features while keeping useful original and engineered features.

### 8.1 Remove `Id`

`Id` is an identifier rather than a meaningful house characteristic, so it was removed.

---

### 8.2 Correlation Analysis

Correlation between numerical features and `SalePrice` was examined.

Correlation was also used to identify highly correlated features that could contain redundant information.

---

### 8.3 Highly Correlated Features

The following highly correlated feature pairs/groups were identified:

- `YearBuilt` ↔ `HouseAge` → **-0.999**
- `YearRemodAdd` ↔ `YearsSinceRemod` → **-0.998**
- `BsmtFinSF1` ↔ `TotalBsmtFinSF` → **0.944**
- `TotalBsmtSF` ↔ `1stFlrSF` → **0.827**
- `TotalBsmtSF` ↔ `TotalSF` → **0.827**
- `TotalBsmtSF` ↔ `TotalLivingArea` → **0.823**
- `1stFlrSF` ↔ `TotalSF` → **0.801**
- `GrLivArea` ↔ `TotRmsAbvGrd` → **0.821**
- `GrLivArea` ↔ `TotalSF` → **0.872**
- `GrLivArea` ↔ `TotalLivingArea` → **0.878**
- `GarageCars` ↔ `GarageArea` → **0.883**
- `TotalSF` ↔ `TotalLivingArea` → **0.998**

---

### 8.4 Remove Redundant Features

The following redundant features were removed:

- `YearBuilt`
- `YearRemodAdd`
- `BsmtFinSF1`
- `TotalLivingArea`
- `GarageArea`

The useful original features and engineered features were retained.

---

### 8.5 Final Feature Selection Check

After feature selection, the training and testing datasets were checked to ensure that they contained the same predictor features.

The feature set was kept consistent between training and testing data.

---

## 9. Categorical Encoding

The remaining categorical features were separated into ordinal and nominal categories before encoding.

### 9.1 Ordinal and Nominal Classification

There were **43 categorical features** in total:

- **18 ordinal features**
- **25 nominal features**

Therefore:

`18 + 25 = 43`

#### Ordinal Features

`ExterQual`, `ExterCond`, `BsmtQual`, `BsmtCond`, `HeatingQC`, `KitchenQual`, `FireplaceQu`, `GarageQual`, `GarageCond`, `BsmtExposure`, `BsmtFinType1`, `BsmtFinType2`, `GarageFinish`, `Functional`, `PavedDrive`, `LotShape`, `LandSlope`, `PoolQC`

#### Nominal Features

`MSZoning`, `Street`, `Alley`, `LandContour`, `LotConfig`, `Neighborhood`, `Condition1`, `Condition2`, `BldgType`, `HouseStyle`, `RoofStyle`, `RoofMatl`, `Exterior1st`, `Exterior2nd`, `MasVnrType`, `Foundation`, `Heating`, `CentralAir`, `Electrical`, `GarageType`, `SaleType`, `SaleCondition`, `MiscFeature`, `Fence`, `Utilities`

---

### 9.2 Ordinal Encoding

Ordinal encoding was used for the 18 ordinal features because their categories have a meaningful order.

For example:

`ExterQual: Po < Fa < TA < Gd < Ex`

The category order was explicitly defined before encoding.

`fit_transform()` was applied to the training data, while `transform()` was applied to the testing data.

---

### 9.3 One-Hot Encoding

One-hot encoding was used for the 25 nominal features because their categories do not have a meaningful numerical order.

`handle_unknown='ignore'` was used so that unseen categories in the test data do not cause an encoding error.

The encoder was fitted only on the training data and then applied to the testing data.

---

### 9.4 Combine Encoded Features

The following were combined:

- Numerical features
- Ordinal encoded features
- One-hot encoded nominal features

This produced the final numerical feature matrices.

---

### 9.5 Encoding Verification

After encoding:

- `X_train` → **(1168, 230)**
- `X_test` → **(292, 230)**

Therefore, the final encoded datasets contain **230 numerical model-input features**.

The increase from the original number of columns is mainly due to one-hot encoding, where one categorical feature can become multiple binary columns.

---

## 10. Feature Scaling

After encoding, all model-input features were numerical.

Standardization was applied using `StandardScaler`.

**Formula:**

`z = (x - mean) / standard deviation`

The scaler was fitted only on the training data and then applied to the testing data.

`X_train → fit_transform()`

`X_test → transform()`

The final scaled datasets were:

- `X_train_scaled` → **(1168, 230)**
- `X_test_scaled` → **(292, 230)**

Scaling changes the feature values but does not change the number of features.

---

## 11. Model Training

Since `SalePrice` is a continuous numerical target, the problem is a **regression problem**.

A Linear Regression model was used as the initial baseline model.

The model was trained using:

`X_train_scaled + y_train`

Predictions were generated using:

`X_test_scaled`

The baseline model provides a reference point for comparing other regression models.

---

## Final Preprocessing Output

### Training Data (`X_train_scaled`)

`X_train_scaled = (1168, 230)`

### Testing Data (`X_test_scaled`)

`X_test_scaled = (292, 230)`

### Target Variables

`y_train = (1168,)`

`y_test = (292,)`

The preprocessing workflow is now complete and the datasets are ready for **model evaluation and comparison**.

# Final Preprocessing Output

After completing preprocessing, feature engineering, feature selection, categorical encoding, and feature scaling, the final datasets used for model training and testing are:

### Training Data (`X_train_scaled`)

`X_train_scaled` contains the processed and scaled features used to train the machine learning models.

**Shape:**

`X_train_scaled = (1168, 230)`

### Testing Data (`X_test_scaled`)

`X_test_scaled` contains the same processed and scaled features for evaluating the trained models.

**Shape:**

`X_test_scaled = (292, 230)`

### Target Variables

The corresponding target values are:

- `y_train` → target values for training
- `y_test` → target values for testing

**Shapes:**

`y_train = (1168,)`

`y_test = (292,)`

### Final Model Training Data

The model is trained using:

`X_train_scaled + y_train`

and evaluated using:

`X_test_scaled + y_test`

### Final Dataset Structure

```text
X_train_scaled (1168, 230) + y_train (1168,)
                         ↓
                   Model Training
                         ↓
              Trained Machine Learning Model
                         ↓
X_test_scaled (292, 230) + y_test (292,)
                         ↓
                  Model Evaluation

**Preprocessed by:** Jayashan Manodya  
**Student ID:** IT24102348